src/operator/controllers/secrets/manager.go

```diff
@@ -258,3 +288,97 @@ func (m *KubernetesSecretsManager) isUpdateNeeded(existingSecretConfig secretsty
+	})
+	if err != nil {
+		return err
+	}
+	// create unique owner list
+	owners := make(map[secretstypes.PodOwnerIdentifier]client.Object)
+	for _, pod := range podList.Items {
+		if _, ok := pod.Annotations[metadata.ShouldRestartOnRenewalAnnotation]; ok {
+			owner, err := m.serviceIdResolver.GetOwnerObject(ctx, &pod)
+			if err != nil {
+				return err
+			}
+			owners[secretstypes.PodOwnerIdentifier{Name: owner.GetName(), GroupVersionKind: owner.GetObjectKind().GroupVersionKind()}] = owner
+		}
+	}
+	for _, owner := range owners {
+		logrus.Infof("Restarting pods for owner %s of type %s after certificate renewal",
+			owner.GetName(), owner.GetObjectKind().GroupVersionKind().Kind)
+		err = m.TriggerPodRestarts(ctx, owner, secret)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// TriggerPodRestarts edits the pod owner's template spec with an annotation about the secret's expiry date
+// If the secret is refreshed, its expiry will be updated in the pod owner's spec which will trigger the pods to restart
+func (m *KubernetesSecretsManager) TriggerPodRestarts(ctx context.Context, owner client.Object, secret *corev1.Secret) error {
+	var err error
+	kind := owner.GetObjectKind().GroupVersionKind().Kind
+	switch kind {
+	case "Deployment":
+		deployment := v1.Deployment{}
+		if err := m.Get(ctx, types.NamespacedName{Namespace: secret.Namespace, Name: owner.GetName()}, &deployment); err != nil {
+			return err
+		}
+		deployment.Spec.Template = m.updatePodTemplateSpec(deployment.Spec.Template)
+		err = m.Update(ctx, &deployment)
+		m.eventRecorder.Eventf(&deployment, corev1.EventTypeNormal, CertRenewReason, "Successfully restarted Deployment after secret '%s' renewal", secret.Name)
+	case "ReplicaSet":
+		replicaSet := v1.ReplicaSet{}
+		if err := m.Get(ctx, types.NamespacedName{Namespace: secret.Namespace, Name: owner.GetName()}, &replicaSet); err != nil {
+			return err
+		}
+		replicaSet.Spec.Template = m.updatePodTemplateSpec(replicaSet.Spec.Template)
+		err = m.Update(ctx, &replicaSet)
+		m.eventRecorder.Eventf(&replicaSet, corev1.EventTypeNormal, CertRenewReason, "Successfully restarted ReplicaSet after secret '%s' renewal", secret.Name)
+	case "StatefulSet":
+		statefulSet := v1.StatefulSet{}
+		if err := m.Get(ctx, types.NamespacedName{Namespace: secret.Namespace, Name: owner.GetName()}, &statefulSet); err != nil {
+			return err
+		}
+		statefulSet.Spec.Template = m.updatePodTemplateSpec(statefulSet.Spec.Template)
+		err = m.Update(ctx, &statefulSet)
+		m.eventRecorder.Eventf(&statefulSet, corev1.EventTypeNormal, CertRenewReason, "Successfully restarted StatefulSet after secret '%s' renewal", secret.Name)
+
+	case "DaemonSet":
+		daemonSet := v1.DaemonSet{}
+		if err := m.Get(ctx, types.NamespacedName{Namespace: secret.Namespace, Name: owner.GetName()}, &daemonSet); err != nil {
+			return err
+		}
+		daemonSet.Spec.Template = m.updatePodTemplateSpec(daemonSet.Spec.Template)
+		err = m.Update(ctx, &daemonSet)
```

Comment:
Please check the error right after this line - checking it later is prone to having it ignored as someone changes the code later. If they overwrite err, they won't realize they didn't check this error, since errors are normally checked immediately after, and this breaks with that pattern.